src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp

```diff
@@ -0,0 +1,143 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2021 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+
+#include "windowsplatformtheme.h"
+#include "log.h"
+#include <Windows.h>
+
+using namespace mu::ui;
+using namespace mu::async;
+
+using ThemeType = IUiConfiguration::ThemeType;
+
+static const std::wstring windowsThemeKey = L"Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize";
+static const std::wstring windowsThemeSubkey = L"AppsUseLightTheme";
+
+HKEY hKey = nullptr;
+
+WindowsPlatformTheme::WindowsPlatformTheme()
+{
+    using fnRtlGetNtVersionNumbers = void(WINAPI*)(LPDWORD major, LPDWORD minor, LPDWORD build);
+    auto rtlGetNtVersionNumbers
+        = reinterpret_cast<fnRtlGetNtVersionNumbers>(GetProcAddress(GetModuleHandleW(L"ntdll.dll"), "RtlGetNtVersionNumbers"));
+    if (rtlGetNtVersionNumbers) {
+        DWORD major, minor, buildNumber;
+        rtlGetNtVersionNumbers(&major, &minor, &buildNumber);
+        m_buildNumber = buildNumber & ~0xF0000000;
+    }
+}
+
+WindowsPlatformTheme::~WindowsPlatformTheme()
+{
+    stopListening();
+}
+
+void WindowsPlatformTheme::init()
+{
+    updateListeningStatus(configuration()->preferredThemeType());
+
+    configuration()->preferredThemeTypeChanged().onReceive(this, [this](const ThemeType& newThemeType) {
+        updateListeningStatus(newThemeType);
+    });
+}
+
+void WindowsPlatformTheme::updateListeningStatus(ThemeType themeType)
+{
+    if (themeType == ThemeType::FOLLOW_SYSTEM_THEME) {
+        startListening();
+    } else {
+        stopListening();
+    }
+}
+
+void WindowsPlatformTheme::startListening()
+{
+    if (m_isListening) {
+        return;
+    }
+    m_isListening = true;
+    if (RegOpenKeyEx(HKEY_CURRENT_USER, windowsThemeKey.c_str(), 0,
+                     KEY_NOTIFY | KEY_CREATE_SUB_KEY | KEY_ENUMERATE_SUB_KEYS | KEY_QUERY_VALUE | KEY_WOW64_64KEY,
+                     &hKey) == ERROR_SUCCESS) {
+        std::thread listenThread([this]() { th_listen(); });
+        listenThread.detach();
```

Comment:
This moment can lead to a potential crash - if, at the time of closing the application when the WindowsPlatformTheme object is destroyed, a call to its class member (`m_darkModeSwitched`). In this case, the probability of failure is extremely small.
But in general, it is better to do this in such cases:
1. Make the thread a member of the class
2. In the `stopListening` method, do this

```
void WindowsPlatformTheme::stopListening()
{
    m_isListening = false;
    m_listenThread.join();
}
```

Reply:
But if I would put m_listenThread.join() in stopListening, then it blocks the app each time when you disable automatic theme switching, for at most 4000 ms (or whatever time we set in WaitForSingleObject). Maybe another option would be to call it only in the destructor, but I think that could theoretically still cause problems.

But, I just found a potentially very useful StackOverflow post (https://stackoverflow.com/a/6033171/11432719), so I will try that.

Reply:
My plan worked: now a "Stop" event is sent, so that `WaitForMultipleObjects` will return directly when it receives the event and the thread finishes (almost) directly, so calling `join()` won't cause waiting any more. Now it should be a bit safer!

Reply:
Well done!